### PR TITLE
Add config flags: skipRows, numRows, ignoreEmptyLines, allowMissingColumns, ignoreExcessColumns.

### PR DESCRIPTION
**Pull request by kosak**
Also allow the nullValue literal to be set to null (which means "unset").

### CODE REVIEW

**Review by jcferretti — APPROVED** [review]

File: src/main/java/io/deephaven/csv/reading/ParseInputToDenseStorage.java [hunk 1]
```diff
@@ -37,106 +37,199 @@
...
-        // treat
-        // it as if the missing cells contained the nullValueLiteral.
-        final byte[] nullValueBytes = nullValueLiteral.getBytes(StandardCharsets.UTF_8);
-        final ByteSlice nullSlice = new ByteSlice(nullValueBytes, 0, nullValueBytes.length);
+        if (skipRows < 0) {
+            throw new IllegalArgumentException("skipRows = " + skipRows);
+        }
+        if (numRows < 0) {
+            throw new IllegalArgumentException("numRows = " + numRows);
+        }
```
[the remark this answers is not in the document]
Comment (kosak) [reply src/main/java/io/deephaven/csv/reading/ParseInputToDenseStorage.java]: Done - please take another look

File: src/main/java/io/deephaven/csv/reading/CsvReader.java [hunk 2]
```diff
@@ -112,7 +112,10 @@ public static Result read(final CsvSpecs specs, final InputStream stream, final
         try {
             final Future<Long> numRowsFuture =
                     exec.submit(
-                            () -> ParseInputToDenseStorage.doit(firstDataRow, specs.nullValueLiteral(), grabber, dsws));
+                            () -> ParseInputToDenseStorage.doit(firstDataRow, specs.nullValueLiteral(),
+                                    specs.skipRows(), specs.numRows(), grabber, dsws,
+                                    specs.ignoreEmptyLines(), specs.allowMissingColumns(),
+                                    specs.ignoreExcessColumns()));
```
[the remark this answers is not in the document]
Comment (kosak) [reply src/main/java/io/deephaven/csv/reading/CsvReader.java]: Done - please take another look

File: src/main/java/io/deephaven/csv/CsvSpecs.java [hunk 3]
```diff
@@ -170,6 +205,24 @@ public static Builder builder() {
         return ImmutableCsvSpecs.builder();
     }
 
+    /**
+     * Validates the {@link CsvSpecs}.
+     */
+    @Check
+    protected void check() {
+        // To be friendly, we report all the problems we find at once.
```
Comment (devinrsmith) [inline src/main/java/io/deephaven/csv/CsvSpecs.java:213]: I think immutables library will be friendly for you (might want to double check) if you have a `@Check` per precondition.

File: src/main/java/io/deephaven/csv/CsvSpecs.java [hunk 4]
```diff
@@ -170,6 +205,24 @@ public static Builder builder() {
         return ImmutableCsvSpecs.builder();
     }
 
+    /**
+     * Validates the {@link CsvSpecs}.
+     */
+    @Check
+    protected void check() {
```
Comment (devinrsmith) [inline src/main/java/io/deephaven/csv/CsvSpecs.java:212]: I don't think this should have the protected modifier (wider than it needs to be); should be no modifier.

File: src/main/java/io/deephaven/csv/CsvSpecs.java (hunk 3 above)
Comment (kosak) [reply src/main/java/io/deephaven/csv/CsvSpecs.java:213]: Nope: if I write code like this

```
    @Check
    void check1() {
        throw new RuntimeException("I am sad");
    }

    @Check
    void check2() {
        throw new RuntimeException("I am very sad");
    }

    @Check
    void check3() {
        throw new RuntimeException("I am really sad");
    }
```

Immutables writes code like this. Fails on the first failing check (it happens to reverse the order), doesn't make any attempt to gather them together.

```
  private static ImmutableCsvSpecs validate(ImmutableCsvSpecs instance) {
    instance.check3();
    instance.check2();
    instance.check1();
    return instance;
  }
```

File: src/main/java/io/deephaven/csv/CsvSpecs.java (hunk 4 above)
Comment (kosak) [reply src/main/java/io/deephaven/csv/CsvSpecs.java]: Changed. However, be aware I was copying the style I saw at https://immutables.github.io/immutable.html#precondition-check-method , which does use "protected"

**Pull request merged**